code/modules/mob/new_player/login.dm

```diff
@@ -11,27 +18,223 @@
 		forceMove(locate(1,1,1))
 	lastarea = get_area(src.loc)
 
-	sight |= SEE_TURFS
+	initialize_lobby_screen()
+
+	GLOB.new_player_list += src
 
 	. = ..()
 
-	new_player_panel()
 	addtimer(CALLBACK(src, PROC_REF(lobby)), 4 SECONDS)
 
+/mob/new_player/proc/initialize_lobby_screen()
+	if(!client)
+		return
+
+	var/datum/tgui/ui = SStgui.get_open_ui(src, src)
+	if(ui)
+		ui.close()
+
+	winset(src, "lobby_browser", "is-disabled=false;is-visible=true")
+	lobby_window = new(client, "lobby_browser")
+	lobby_window.initialize(
+		assets = list(
+				get_asset_datum(/datum/asset/simple/tgui),
+			)
+	)
+
+	tgui_interact(src)
+
+/mob/new_player/tgui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "LobbyMenu", window = lobby_window)
+		ui.closeable = FALSE
+		ui.open(preinitialized = TRUE)
+
+/mob/new_player/ui_state(mob/user)
+	return GLOB.always_state
+
+/mob/new_player/ui_data(mob/user)
+	. = ..()
+
+	.["character_name"] = client.prefs ? client.prefs.real_name : client.key
+
+	var/postfix_text = (client.xeno_postfix) ? ("-"+client.xeno_postfix) : ""
+	var/prefix_text = (client.xeno_prefix) ? client.xeno_prefix : "XX"
+	.["xeno_prefix"] = prefix_text
+	.["xeno_postfix"] = postfix_text
+
+	.["tutorials_ready"] = SSticker?.current_state == GAME_STATE_PLAYING
+	.["round_start"] = !SSticker || !SSticker.mode || SSticker.current_state <= GAME_STATE_PREGAME
+	.["readied"] = ready
+
+	.["upp_enabled"] = GLOB.master_mode == /datum/game_mode/extended/faction_clash/cm_vs_upp::name
+	.["xenomorph_enabled"] = client.prefs && (client.prefs.get_job_priority(JOB_XENOMORPH))
+	.["predator_enabled"] = SSticker.mode?.flags_round_type & MODE_PREDATOR && SSticker.mode.check_predator_late_join(src, FALSE)
+	.["fax_responder_enabled"] = SSticker.mode?.check_fax_responder_late_join(src, FALSE)
+
+	.["lobby_author"] = SSlobby_art.author
```

Comment:
data that doesn't change should probably go into ui_static_data

Reply:
if it does end up changing but only very occasionally, you can manually trigger a full update on clients (i.e. when the lobby art is loaded)